### ISSUE TITLE
The GLSL findMSB function does not seem to work after translation to Metal Shading Language

### ISSUE THREAD

**Issue by elmar-k** [KhronosGroup/SPIRV-Cross#1413]
Dear SPIRV-Cross team,

I started testing our Vulkan app with MoltenVK in MacOS, and found that the Metal translation of findMSB didn't work.

Here is a GLSL example code:

uint edgelen;

myarray[0]=0x1234;
edgelen=20;
myarray[1]=edgelen;
myarray[2]=findMSB(edgelen);
myarray[3]=findMSB(uint(20));
myarray[4]=uint(log2(20.0));
myarray[5]=0x5678;

Running this in Linux and reading back the result gives me these expected bytes:

00001234 00000014 00000004 00000004 00000004 00005678

In MacOS 10.14 with Geforce GT 750M and MoltenVK from the latest SDK (vulkansdk-macos-1.2.141.2.dmg from June 11, which includes SPIRV-Cross with GitHub hash d385bf096f5dabbc4cdaeb6872b0f64be1a63ad0) I get

00001234 00000014 00000000 00000000 00000004 00005678

Note that findMSB returns 0 instead of 4.

I compiled and ran SPIRV-Cross to look at the Metal shader code, which is:

myarray[0] = 4660u;
myarray[1] = 20u;
uint _267 = uint(int(spvFindUMSB(20u)));
myarray[2] = _267;
myarray[3] = _267;
myarray[4] = 4u;
myarray[5] = 22136u;

So something seems wrong with spvFindUMSB. I tried to help by looking into SPIRV-Cross doing a

git clone https://github.com/KhronosGroup/SPIRV-Cross.git SPIRV-Cross
cd SPIRV-Cross
git checkout d385bf096f5dabbc4cdaeb6872b0f64be1a63ad0

(this was the SPIRV-Version I found I the latest MoltenVK source).

But "unfortunately" the SPIRV-Code looks OK to me:

/ Implementation of the unsigned GLSL findMSB() function
template
inline T spvFindUMSB(T x)
{
return select(clz(T(0)) - (clz(x) + T(1)), T(-1), x == T(0));
}

It also looks like there is a testsuite in SPIRV-Cross which checks this function, so I fear someone with more inside knowledge is needed at this point to figure out why this goes wrong.

To workaround the problem on my side, I'm using this slower replacement now, then my shader works:

#define findMSB(VAL) uint(log2(float(VAL)))

(this wouldn't work if VAL was 0, but I don't have this case ;-)

I keep fingers crossed that this report helps you trace the problem. I already reported it to MoltenVK, who sent me here.

Best regards,
Elmar

**Comment by HansKristian-Work** [comment]
Very curious ... Could this be a driver bug? I can't see anything obviously wrong with the UMSB implementation. Wonder if this has something to do with `clz(T(0))`. MSL spec defines what happens here FWIW.

@cdavis5e @billhollings any ideas?

**Comment by billhollings** [comment]
Hmmm...

I just tested the following Metal shader, hacked from a basic Apple compute sample project:

``` c++
#include <metal_stdlib>
using namespace metal;

// Implementation of the unsigned GLSL findMSB() function
template<typename T>
inline T spvFindUMSB(T x)
{
	return select(clz(T(0)) - (clz(x) + T(1)), T(-1), x == T(0));
}

/// This is a Metal Shading Language (MSL) function equivalent to the add_arrays() C function, used to perform the calculation on a GPU.
kernel void add_arrays(device const float* inA,
                       device const float* inB,
                       device float* result,
                       uint index [[thread_position_in_grid]])
{
    // the for-loop is replaced with a collection of threads, each of which
    // calls this function.
//	result[index] = inA[index] + inB[index];
//	result[index] = spvFindUMSB(uint(20));
	result[index] = uint(int(spvFindUMSB(20u)));
}
```

I get the correct answer of 4. And changing the constant to 33 or 66 also gives the correct values of 5 and 6. It also doesn't seem to be affected by changing how things are passed (as in the line above the active line).

I'm running MacBook Pro 2017, Catalina 10.15.5, and I get the same results on both AMD Radeon Pro 560 and IntelHD Graphics 630. I haven't tried on iOS yet. So...it _might_ be a localized driver bug...but if so it is very localized.

@elmar-k What does the rest of your shader look like, and what is the contextual environment under which you are running your shader? Is it parallel compute, how are you reading the results, what type of device memory, how are you syncing GPU/CPU, etc. I'm just trying to remove possible sync issues, or overwrite issues like we were seeing on the [other shader write/read issue](https://github.com/KhronosGroup/MoltenVK/issues/942) you reported.

**Comment by elmar-k** [comment]
Hi Bill,

many thanks for trying. In the mean time, my Vulkan port made progress and I can now select the GPU I want.

Previously, I reported that findMSB does not work in MacOS 10.14 with Geforce GT 750M.
On the same MacBook Pro, switching to Intel Iris Pro, the same code with findMSB works fine. Since this is the same notebook with the same MoltenVK etc., this seems to be a low level bug in the 'clz' function in Apple's nVIDIA drivers.

The next step would be to send the test case you kindly made to someone with MacOS 10.15 and an nVIDIA GPU to see if it has been fixed in the latest MacOS, and then either close the bug or forward to Apple..  Unfortunately I cannot get hold of such a system myself.

Best regards,
Elmar
